nornir/core/connections.py

```diff
@@ -0,0 +1,58 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, NoReturn, Optional, Union
+
+
+from nornir.core.configuration import Config
+
+
+class ConnectionPlugin(ABC):
+    """
+    Connection plugins have to inherit from this class and provide implementations
+    for both the :meth:`open` and :meth:`close` methods.
+
+    Attributes:
+        connection: Underlying connection. Populated by :meth:`open`.
+        state: Dictionary to hold any data that needs to be shared between
+            the connection plugin and the plugin tasks using this connection.
+    """
+
+    __slots__ = ("connection", "state")
+
+    def __init__(self) -> None:
+        self.connection: Union[Any, UnestablishedConnection] = UnestablishedConnection()
+        self.state: Dict[str, Any] = {}
+
+    @abstractmethod
+    def open(
+        self,
+        hostname: str,
+        username: str,
+        password: str,
+        ssh_port: int,
+        network_api_port: int,
+        operating_system: str,
+        nos: str,
+        connection_options: Optional[Dict[str, Any]] = None,
+        configuration: Optional[Config] = None,
+    ) -> None:
+        """
+        Connect to the device and populate the attribute :attr:`connection` with
+        the underlying connection
+        """
+        pass
+
+    @abstractmethod
+    def close(self) -> None:
+        """Close the connection with the device"""
+        pass
+
+
+class UnestablishedConnection(object):
+    def close(self) -> NoReturn:
+        raise ValueError("Connection not established")
+
+    disconnect = close
+
+
+class Connections(Dict[str, ConnectionPlugin]):
```

Comment:
> I am not sure about the purpose of this class.

The purpose is to have a distinct object in case we want to add methods to it.

> It also inherits from the type Dict, which is not correct.

What would be the correct way? From what I read this seems to be the correct way of doing this (it certainly works as I expected). For instance:

```
class Asd(Dict[str, int]):
    pass


a = Asd()
a["asd"] = "asd"

error: Incompatible types in assignment (expression has type "str", target has type "int")
```

> this class could replace both available_connections and connection instances for Host

Not sure they are the same thing though. For instance, the `register_connection` you mention may make sense in one but not the other but I will look into it as I address some of the other comments.

Reply:
I agree this is the right thing to and I even tried a similar implementation, however, this change is a bit more tricky than it looks as it runs into circular imports if you pre-register the connections shipped with nornir. In any case, this is just a minor OCD fix to the internal API so feel free to ticket it so we can take a closer look but I am not going to address it in this PR.

Reply:
#211